### PR TITLE
feat(admin,router,hermes,grpc-sdk,core): middleware patch support

### PR DESCRIPTION
**Pull request by ChrisPdgn**
This PR introduces the feature of patching the middleware array of a grpc route. This applies for admin and app routes. Middleware order can be changed freely, but removal is allowed only to the module that registered the middleware. 

**What kind of change does this PR introduce?** 

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** 

- [x] Yes
- [ ] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `main` branch
- [ ] When resolving a specific issue, it's referenced in the PR's description  (e.g. `fix #xxx`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature

### CODE REVIEW

File: modules/router/src/Router.ts [hunk 1]
```diff
@@ -147,6 +153,7 @@ export default class ConduitDefaultRouter extends ManagedModule<Config> {
     }
     await this.highAvailability();
     this.updateHealth(HealthCheckStatus.SERVING);
+    this.grpcSdk.createModuleClient('router', process.env.SERVICE_IP!);
```
Comment (kon14) [inline modules/router/src/Router.ts:156]: What would this accomplish? The Router's own `grpc-sdk` shouldn't be communicating with itself through gRPC.

File: modules/router/src/Router.ts [hunk 2]
```diff
@@ -461,4 +470,161 @@ export default class ConduitDefaultRouter extends ManagedModule<Config> {
     }
     ConduitGrpcSdk.Logger.log('Recovered routes');
   }
+
+  async patchMiddleware(
```
Comment (kon14) [inline modules/router/src/Router.ts:474]: Maybe rename this to `patchRouteMiddlewares` or sth that would convey we're actually affecting the route's middlewares.

File: modules/router/src/Router.ts [hunk 3]
```diff
@@ -461,4 +470,161 @@ export default class ConduitDefaultRouter extends ManagedModule<Config> {
...
+  async internalPatchMiddleware(
+    path: string,
+    action: string,
+    middleware: string[],
+    moduleUrl?: string,
+  ) {
+    let injected: string[] = [];
+    let removed: string[] = [];
+    if (!moduleUrl) {
+      moduleUrl = await this.grpcSdk.config.getModuleUrlByName('router').then(r => r.url);
```
Comment (kon14) [inline modules/router/src/Router.ts:507]: Add a comment clarifying this is about administrators modifying client API middlewares.

File: modules/router/src/admin/index.ts [hunk 4]
```diff
@@ -49,6 +49,37 @@ export class AdminHandlers {
       }),
       this.routerAdmin.getMiddlewares.bind(this.routerAdmin),
     );
+    this.routingManager.route(
+      {
+        path: '/router/route-middlewares',
+        action: ConduitRouteActions.GET,
+        description: `Returns the middleware of an app route.`,
+        queryParams: {
+          path: ConduitString.Required,
```
Comment (kon14) [inline modules/router/src/admin/index.ts:58]: This is REST-oriented.
Not saying we could/should change this, but we should probably document it a bit.

As is, this should include the module name in the path.
Does this work for hooks? What's the path for a hook given the order swap going on.

File: modules/router/src/admin/index.ts [hunk 5]
```diff
@@ -49,6 +49,37 @@ export class AdminHandlers {
...
+        },
+      },
+      new ConduitRouteReturnDefinition('GetAppRouteMiddleware', {
+        middleware: [TYPE.String],
+      }),
+      this.routerAdmin.getRouteMiddlewares.bind(this.routerAdmin),
+    );
+    this.routingManager.route(
+      {
+        path: '/router/patch-middleware',
```
Comment (kon14) [inline modules/router/src/admin/index.ts:69]: Either `'/router/patch-middleware'` -> `'/router/patch-middlewares'` or update the getter route. Pick one as long as it's consistent in the external APIs.
Also update descriptions.

File: modules/router/src/admin/router.ts [hunk 6]
```diff
@@ -38,6 +41,32 @@ export class RouterAdmin {
     return Array.from(new Set(response));
   }
 
+  async getRouteMiddlewares(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
```
Comment (kon14) [inline modules/router/src/admin/router.ts:44]: Do these work with hooks?

**Review by kon14 — CHANGES_REQUESTED** [review]

File: modules/router/src/admin/router.ts (hunk 6 above)
Comment (ChrisPdgn) [reply modules/router/src/admin/router.ts:44]: After testing it, yes they do

File: libraries/hermes/src/Router.ts [hunk 7]
```diff
@@ -122,4 +128,45 @@ export abstract class ConduitRouter {
...
+    this._registeredRoutes.set(key, route);
+    const routes: { action: string; path: string }[] = [];
+    for (const conduitRoute of this._registeredRoutes.values()) {
+      routes.push({ action: conduitRoute.input.action, path: conduitRoute.input.path });
+    }
+    this.cleanupRoutes(routes);
+  }
+
+  filterMiddlewaresPatch(
+    routeMiddleware: string[],
```
Comment (kon14) [inline libraries/hermes/src/Router.ts:145]: `routeMiddlewares` / `patchMiddlewares`

File: libraries/hermes/src/index.ts [hunk 8]
```diff
@@ -157,11 +158,35 @@ export class ConduitRoutingController {
...
+    this._proxyRouter?.registerMiddleware(middleware, moduleUrl);
+  }
+
+  patchRouteMiddlewares(patch: MiddlewarePatch) {
+    this._restRouter?.patchRouteMiddlewares(patch);
+    this._graphQLRouter?.patchRouteMiddlewares(patch);
+  }
+
+  filterMiddlewaresPatch(
+    routeMiddleware: string[],
```
Comment (kon14) [inline libraries/hermes/src/index.ts:174]: `routeMiddlewares` / `patchMiddlewares`

**Review by kon14 — CHANGES_REQUESTED** [review]
Just a couple leftovers.

**Review by kon14 — APPROVED** [review]

**Pull request merged**